### PR TITLE
Fix layout handling in CmdBeginRenderPass, CmdNextSubpass

### PR DESCRIPTION
**Pull request by chrisforbes**
Contains fixes for issues #704, #655, #716

### CODE REVIEW

File: tests/layer_validation_tests.cpp [hunk 1]
```diff
@@ -3010,6 +3010,258 @@ TEST_F(VkLayerTest, BindMemoryToDestroyedObject) {
 
 #if DRAW_STATE_TESTS
 
+TEST_F(VkLayerTest, RenderPassInitialLayoutUndefined) {
+    TEST_DESCRIPTION("Ensure that CmdBeginRenderPass with an attachment's initialLayout"
```
Comment (tobine) [inline tests/layer_validation_tests.cpp]: Need a space at the end of this line. Also, did you run clang-format on the test changes? Test file has smaller column width that appears to be violated here and/or on line 3106 below.

File: tests/layer_validation_tests.cpp [hunk 2]
```diff
@@ -3010,6 +3010,258 @@ TEST_F(VkLayerTest, BindMemoryToDestroyedObject) {
...
+    VkFramebufferCreateInfo fci = {
+        VK_STRUCTURE_TYPE_FRAMEBUFFER_CREATE_INFO, nullptr,
+        0, rp, 1, &view,
+        32, 32, 1
+    };
+    VkFramebuffer fb;
+    err = vkCreateFramebuffer(m_device->device(), &fci, nullptr, &fb);
+    ASSERT_VK_SUCCESS(err);
+
+    // Record a single command buffer which uses this renderpass twice. The
```
Comment (tobine) [inline tests/layer_validation_tests.cpp]: Is this comment applicable?

File: tests/layer_validation_tests.cpp [hunk 3]
```diff
@@ -3010,6 +3010,258 @@ TEST_F(VkLayerTest, BindMemoryToDestroyedObject) {
...
+    VkFramebufferCreateInfo fci = {
+        VK_STRUCTURE_TYPE_FRAMEBUFFER_CREATE_INFO, nullptr,
+        0, rp, 0, nullptr,
+        32, 32, 1
+    };
+    VkFramebuffer fb;
+    err = vkCreateFramebuffer(m_device->device(), &fci, nullptr, &fb);
+    ASSERT_VK_SUCCESS(err);
+
+    // Record a single command buffer which uses this renderpass twice. The
```
Comment (tobine) [inline tests/layer_validation_tests.cpp]: Is this comment applicable?

**Comment by tobine** [comment]
Potentially a few minor clean-ups in the tests, otherwise
LGTM :+1:

File: tests/layer_validation_tests.cpp (hunk 2 above)
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: No, leftover. Both bogus instances replaced locally with comments describing what they are actually trying to do.

**Comment by chrisforbes** [comment]
Feedback incorporated; added fixes for another couple of issues uncovered by the last test.

**Comment by chrisforbes** [comment]
Picked to master.

**Pull request merged**